spark/src/test/scala/org/apache/comet/CometCastSuite.scala

```diff
@@ -176,28 +697,38 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
-
-        if (CometSparkSessionExtensions.isSpark34Plus) {
-          // We have to workaround https://github.com/apache/datafusion-comet/issues/293 here by
-          // removing the "Execution error: " error message prefix that is added by DataFusion
-          val cometMessage = actual.getMessage
-            .substring("Execution error: ".length)
-
-          assert(expected.getMessage == cometMessage)
-        } else {
-          // Spark 3.2 and 3.3 have a different error message format so we can't do a direct
-          // comparison between Spark and Comet.
-          // Spark message is in format `invalid input syntax for type TYPE: VALUE`
-          // Comet message is in format `The value 'VALUE' of the type FROM_TYPE cannot be cast to TO_TYPE`
-          // We just check that the comet message contains the same invalid value as the Spark message
-          val sparkInvalidValue =
-            expected.getMessage.substring(expected.getMessage.indexOf(':') + 2)
-          assert(actual.getMessage.contains(sparkInvalidValue))
+        checkSparkMaybeThrows(df) match {
+          case (None, None) =>
+          // neither system threw an exception
+          case (None, Some(e)) =>
+            // Spark succeeded but Comet failed
+            throw e
+          case (Some(e), None) =>
+            // Spark failed but Comet succeeded
+            fail(s"Comet should have failed with ${e.getCause.getMessage}")
+          case (Some(sparkException), Some(cometException)) =>
+            // both systems threw an exception so we make sure they are the same
+            val sparkMessage = sparkException.getCause.getMessage
+            // We have to workaround https://github.com/apache/datafusion-comet/issues/293 here by
+            // removing the "Execution error: " error message prefix that is added by DataFusion
+            val cometMessage = cometException.getCause.getMessage
+              .replace("Execution error: ", "")
+            if (CometSparkSessionExtensions.isSpark34Plus) {
+              assert(cometMessage == sparkMessage)
+            } else {
+              // Spark 3.2 and 3.3 have a different error message format so we can't do a direct
+              // comparison between Spark and Comet.
+              // Spark message is in format `invalid input syntax for type TYPE: VALUE`
+              // Comet message is in format `The value 'VALUE' of the type FROM_TYPE cannot be cast to TO_TYPE`
+              // We just check that the comet message contains the same invalid value as the Spark message
+              val sparkInvalidValue = sparkMessage.substring(sparkMessage.indexOf(':') + 2)
+              assert(cometMessage.contains(sparkInvalidValue))
+            }
         }
 
         // try_cast() should always return null for invalid inputs
-        val df2 = spark.sql(s"select try_cast(a as ${toType.sql}) from t")
+        val df2 =
+          spark.sql(s"select a, try_cast(a as ${toType.sql}) from t order by a")
```

Comment:
what if a is null, the try_cast return null and will be treated as invalid input?

Reply:
try_cast should return null for null inputs and also for invalid inputs

Reply:
Right, I mean. If there is a test to cast an int to string and this is supported and valid, but I have a NULL in my dataset, so try_cast returns NULL as well. will that be treated as false negative?

Reply:
If the input dataset has a mix of valid, invalid, and null values, we shoul expect to see try_cast produce null for the rows that are null or invalid. The test is checking that we produce the same results as Spark, so I think we have everything covered, but maybe I am missing something?

Reply:
I'd like to merge this one so that we can start rebasing the other pending CAST PRs, but let's continue the conversation.